pkgs/build-support/rust/fetch-cargo-vendor-util.py

```diff
@@ -0,0 +1,247 @@
+
+
+def download_git_tree(url: str, git_sha_rev: str, out_dir: str):
+
+    tree_out_dir = os.path.join(out_dir, "git", git_sha_rev)
+    eprint(f"Fetching {url}#{git_sha_rev} -> git/{git_sha_rev}")
+
+    subprocess.check_output(["nix-prefetch-git", "--builder", "--quiet", "--url", url, "--rev", git_sha_rev, "--out", tree_out_dir])
+
+
+GIT_SOURCE_REGEX = re.compile("git\\+(?P<url>[^?]+)(\\?(?P<type>rev|tag|branch)=(?P<value>.*))?#(?P<git_sha_rev>.*)")
+
+
+def parse_git_source(source: str):
+    matches = GIT_SOURCE_REGEX.match(source)
+    if matches is None:
+        raise Exception(f"Unable to process git source: {source}.")
+    return matches.groupdict()
+
+
+def create_vendor_staging(lockfile_path: str, out_dir: str):
+    cargo_toml = load_toml(lockfile_path)
+
+    git_packages: list[dict] = []
+    registry_packages: list[dict] = []
+
+    for pkg in cargo_toml["package"]:
+        # ignore local dependenices
+        if "source" not in pkg.keys():
+            eprint(f"Skipping local dependency: {pkg["name"]}")
+            continue
+        source = pkg["source"]
+
+        if source.startswith("git+"):
+            git_packages.append(pkg)
+        elif source.startswith("registry+"):
+            registry_packages.append(pkg)
+        else:
+            raise Exception(f"Can't process source: {source}.")
+
+    git_sha_rev_to_url = {}
+    for pkg in git_packages:
+        source_info = parse_git_source(pkg["source"])
+        git_sha_rev_to_url[source_info["git_sha_rev"]] = source_info["url"]
+
+    os.makedirs(out_dir, exist_ok=True)
+    shutil.copy(lockfile_path, os.path.join(out_dir, "Cargo.lock"))
+
+    pool = mp.Pool(min(10, mp.cpu_count()))
+
+    if len(git_packages) != 0:
+        os.mkdir(os.path.join(out_dir, "git"))
+        # run download jobs in parallel
+        pool.starmap(download_git_tree, ((url, git_sha_rev, out_dir) for git_sha_rev, url in git_sha_rev_to_url.items()))
+
+    if len(registry_packages) != 0:
+        os.mkdir(os.path.join(out_dir, "tarballs"))
+        # run download jobs in parallel
+        pool.starmap(download_tarball, ((pkg, out_dir) for pkg in registry_packages))
+
+    pool.close()
+
+
+def get_manifest_metadata(manifest_path: str):
+    output = subprocess.check_output(["cargo", "metadata", "--format-version", "1", "--no-deps", "--manifest-path", manifest_path])
+    return json.loads(output)
+
+
+def try_get_crate_manifest_path_from_mainfest_path(manifest_path: str, crate_name: str) -> str | None:
+    metadata = get_manifest_metadata(manifest_path)
+    crate_manifest_path = next((pkg["manifest_path"] for pkg in metadata["packages"] if pkg["name"] == crate_name), None)
+    return crate_manifest_path
+
+
+def find_crate_manifest_in_tree(tree: str, crate_name: str):
+    # in some cases Cargo.toml is not located at the top level, so we also look at subdirectories
+    manifest_paths = (os.path.join(tree, path) for path in glob.iglob("**/Cargo.toml", root_dir=tree, recursive=True))
```

Comment:
Consider using `pathlib` here.

Reply:
It adds some type safety and makes joining path a bit nicer and more portable.

Reply:
Refactored the entire thing to use pathlib.